File: .skills/openclaw-skills/skills/robertstarry-gif/liuzln-openclaw-skills-wechat-article-fetcher/examples/fetch_wechat_article.py
#!/usr/bin/env python3
"""
微信公众号文章爬取脚本（增强版）

专门用于爬取微信公众号文章内容，支持保存图片。
"""

import asyncio
import json
import logging
import os
import re
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any
from urllib.parse import urlparse, unquote

try:
    from playwright.async_api import async_playwright, Page
except ImportError:
    print("请先安装 Playwright: pip install playwright")
    print("然后安装浏览器: playwright install chromium")
    exit(1)


# 配置日志
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


# 配置
CONFIG = {
    "headless": True,
    "timeout": 60000,
    "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "url": "https://mp.weixin.qq.com/s/HTGvy5C6SYyr5XQhTfTfzw",
    "save_images": True,  # 是否保存图片
    "images_dir": "images",  # 图片保存目录
    "image_quality": 90  # 图片质量 (0-100)
}


def sanitize_filename(filename: str) -> str:
    """清理文件名，移除非法字符"""
    filename = re.sub(r'[<>:"/\\|?*]', '_', filename)
    filename = filename.strip()
    if len(filename) > 100:
        filename = filename[:100]
    return filename


async def download_image(page: Page, img_url: str, save_path: Path) -> bool:
    """下载图片"""
    try:
        # 使用 page.request 下载图片
        response = await page.request.get(img_url)
        if response.ok:
            image_data = await response.body()
            with open(save_path, 'wb') as f:
                f.write(image_data)
            logger.info(f"✅ 已保存图片: {save_path}")
            return True
        else:
            logger.warning(f"❌ 下载图片失败 (HTTP {response.status}): {img_url}")
            return False
    except Exception as e:
        logger.warning(f"❌ 下载图片异常: {img_url} - {e}")
        return False


async def extract_and_save_images(page: Page, images_dir: Path) -> List[Dict[str, Any]]:
    """提取并保存文章中的图片"""
    images_info = []
    
    try:
        # 获取所有图片元素
        img_elements = await page.query_selector_all("#js_content img, .rich_media_content img")
        
        if not img_elements:
            logger.info("未找到图片元素")
            return images_info
        
        logger.info(f"找到 {len(img_elements)} 张图片")
        
        for idx, img_element in enumerate(img_elements, 1):
            try:
                # 获取图片 URL
                src = await img_element.get_attribute("src")
                data_src = await img_element.get_attribute("data-src")
                
                # 优先使用 data-src（微信公众号常用）
                img_url = data_src or src
                
                if not img_url:
                    continue
                
                # 确保 URL 完整
                if img_url.startswith("//"):
                    img_url = "https:" + img_url
                elif img_url.startswith("/"):
                    img_url = "https://mp.weixin.qq.com" + img_url
                
                # 获取图片 alt 作为文件名
                alt = await img_element.get_attribute("alt") or f"image_{idx}"
                alt = sanitize_filename(alt)
                
                # 确定文件扩展名
                parsed = urlparse(img_url)
                path = unquote(parsed.path)
                ext = Path(path).suffix.lower()
                if not ext or ext not in ['.jpg', '.jpeg', '.png', '.gif', '.webp']:
                    ext = '.jpg'
                
                # 保存图片
                img_filename = f"{idx:03d}_{alt}{ext}"
                img_path = images_dir / img_filename
                
                success = await download_image(page, img_url, img_path)
                
                images_info.append({
                    "index": idx,
                    "url": img_url,
                    "alt": alt,
                    "filename": img_filename,
                    "success": success
                })
                
            except Exception as e:
                logger.warning(f"处理图片 {idx} 时出错: {e}")
                continue
        
    except Exception as e:
        logger.error(f"提取图片时出错: {e}", exc_info=True)
    
    return images_info


async def fetch_wechat_article(url: str = None):
    """爬取微信公众号文章"""
    if url is None:
        url = CONFIG["url"]
    
    async with async_playwright() as p:
        # 启动浏览器
        browser = await p.chromium.launch(
            headless=CONFIG["headless"],
            args=["--no-sandbox", "--disable-dev-shm-usage"]
        )
        page = await browser.new_page(
            user_agent=CONFIG["user_agent"],
            viewport={"width": 1920, "height": 1080}
        )
        page.set_default_timeout(CONFIG["timeout"])
        
        try:
            logger.info(f"正在访问微信公众号文章: {url}")
            
            # 访问页面
            try:
                await page.goto(url, wait_until="domcontentloaded")
            except Exception as e:
                logger.warning(f"domcontentloaded 超时，尝试 load: {e}")
                await page.goto(url, wait_until="load")
            
            # 等待页面渲染
            await asyncio.sleep(5)
            
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            
            # 创建目录
            screenshots_dir = Path("screenshots")
            results_dir = Path("results")
            images_dir = Path(CONFIG["images_dir"]) / timestamp
            
            screenshots_dir.mkdir(exist_ok=True)
            results_dir.mkdir(exist_ok=True)
            if CONFIG["save_images"]:
                images_dir.mkdir(parents=True, exist_ok=True)
            
            # 截图
            try:
                screenshot_file = screenshots_dir / f"wechat_article_{timestamp}.png"
                await page.screenshot(path=str(screenshot_file), full_page=True)
                logger.info(f"已保存完整截图: {screenshot_file}")
            except Exception as e:
                logger.warning(f"截图失败: {e}")
            
            # 提取文章标题
            title = ""
            title_selectors = [
                "#activity-name",
                ".rich_media_title",
                "h1",
                "title"
            ]
            
            for selector in title_selectors:
                try:
                    element = await page.query_selector(selector)
                    if element:
                        text = await element.inner_text()
                        if text and len(text.strip()) > 5:
                            title = text.strip()
                            logger.info(f"找到标题 (选择器: {selector}): {title}")
                            break
                except Exception as e:
                    logger.debug(f"选择器 {selector} 失败: {e}")
                    continue
            
            # 提取作者
            author = ""
            author_selectors = [
                "#js_name",
                ".rich_media_meta_nickname",
                ".profile_nickname"
            ]
            
            for selector in author_selectors:
                try:
                    element = await page.query_selector(selector)
                    if element:
                        text = await element.inner_text()
                        if text:
                            author = text.strip()
                            logger.info(f"找到作者 (选择器: {selector}): {author}")
                            break
                except Exception as e:
                    logger.debug(f"选择器 {selector} 失败: {e}")
                    continue
            
            # 提取发布时间
            publish_date = ""
            date_selectors = [
                "#publish_time",
                ".rich_media_meta_text",
                "em#post-date"
            ]
            
            for selector in date_selectors:
                try:
                    element = await page.query_selector(selector)
                    if element:
                        text = await element.inner_text()
                        if text:
                            publish_date = text.strip()
                            logger.info(f"找到时间 (选择器: {selector}): {publish_date}")
                            break
                except Exception as e:
                    logger.debug(f"选择器 {selector} 失败: {e}")
                    continue
            
            # 提取文章内容
            content = ""
            content_selectors = [
                "#js_content",
                ".rich_media_content",
                "#img-content"
            ]
            
            for selector in content_selectors:
                try:
                    element = await page.query_selector(selector)
                    if element:
                        text = await element.inner_text()
                        if text and len(text) > 100:
                            content = text
                            logger.info(f"找到内容 (选择器: {selector})，长度: {len(content)} 字符")
                            break
                except Exception as e:
                    logger.debug(f"选择器 {selector} 失败: {e}")
                    continue
            
            if not content:
                try:
                    content = await page.inner_text("body")
                    logger.info("使用 body 内容")
                except Exception as e:
                    logger.warning(f"获取 body 失败: {e}")
                    content = "无法提取内容"
            
            # 提取并保存图片
            images_info = []
            if CONFIG["save_images"]:
                logger.info("开始提取图片...")
                images_info = await extract_and_save_images(page, images_dir)
            
            # 构造结果
            result = {
                "title": title,
                "author": author,
                "publish_date": publish_date,
                "url": url,
                "content": content,
                "images": images_info,
                "images_count": len([i for i in images_info if i["success"]]),
                "images_dir": str(images_dir) if CONFIG["save_images"] else None,
                "fetch_time": datetime.now().isoformat(),
                "length": len(content)
            }
            
            # 保存结果
            result_file = results_dir / f"wechat_article_{timestamp}.json"
            with open(result_file, 'w', encoding='utf-8') as f:
                json.dump(result, f, ensure_ascii=False, indent=2)
            logger.info(f"已保存结果: {result_file}")
            
            return result
            
        finally:
            await browser.close()


async def main():
    """主函数"""
    print("="*80)
    print("微信公众号文章爬取脚本（增强版）")
    print("="*80)
    
    url = CONFIG["url"]
    print(f"📄 目标文章: {url}")
    print(f"🖼️  保存图片: {'是' if CONFIG['save_images'] else '否'}")
    if CONFIG['save_images']:
        print(f"📁 图片目录: {CONFIG['images_dir']}/")
    print("="*80)
    
    result = await fetch_wechat_article(url)
    
    if result:
        print("\n" + "="*80)
        print(f"📰 标题: {result['title']}")
        if result['author']:
            print(f"✍️  作者: {result['author']}")
        if result['publish_date']:
            print(f"📅 发布时间: {result['publish_date']}")
        print(f"🔗 链接: {result['url']}")
        print(f"📝 内容长度: {result['length']} 字符")
        if CONFIG['save_images']:
            print(f"🖼️  图片保存: {result['images_count']}/{len(result['images'])} 张")
            print(f"📁 图片目录: {result['images_dir']}")
        print("="*80)
        print("\n📖 内容摘要:")
        print(result['content'][:1000] + "..." if len(result['content']) > 1000 else result['content'])
        print("\n" + "="*80)
        print("\n✅ 爬取完成！")
        print(f"📄 JSON 结果: results/wechat_article_*.json")
        print(f"🖼️  完整截图: screenshots/wechat_article_*.png")
        if CONFIG['save_images']:
            print(f"📷 文章图片: {result['images_dir']}/")
    else:
        print("\n❌ 未能获取文章内容")


if __name__ == "__main__":
    asyncio.run(main())

File: .skills/openclaw-skills/skills/adgai115/wechat-article-fetcher-safe/fetch-wechat-article.js
const puppeteer = require('puppeteer-core');
const fs = require('fs');
const path = require('path');
const os = require('os');
const https = require('https');
const http = require('http');
const Tesseract = require('tesseract.js');

/**
 * OCR 识别图片文字
 */
async function recognizeImageText(imagePath, imageIndex, totalImages) {
  try {
    console.log(`  🔍 正在识别图片 ${imageIndex}/${totalImages}...`);
    const { data: { text } } = await Tesseract.recognize(
      imagePath,
      'chi_sim+eng', // 中文 + 英文
      {
        logger: m => {
          if (m.status === 'recognizing text') {
            process.stdout.write(`\r  OCR 进度：${(m.progress * 100).toFixed(0)}%`);
          }
        }
      }
    );
    console.log(''); // 换行
    return text.trim();
  } catch (err) {
    console.error(`  ❌ OCR 失败：${err.message}`);
    return '';
  }
}

/**
 * 下载图片到本地
 */
async function downloadImage(url, outputDir, filename) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http;
    const filepath = path.join(outputDir, filename);
    
    client.get(url, (res) => {
      if (res.statusCode === 302 || res.statusCode === 301) {
        // 重定向
        downloadImage(res.headers.location, outputDir, filename).then(resolve).catch(reject);
        return;
      }
      
      const file = fs.createWriteStream(filepath);
      res.pipe(file);
      
      file.on('finish', () => {
        file.close();
        resolve(filepath);
      });
    }).on('error', (err) => {
      fs.unlink(filepath, () => {});
      reject(err);
    });
  });
}

/**
 * 自动查找 Chrome 路径
 */
function findChromePath() {
  if (process.platform === 'win32') {
    const paths = [
      'C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe',
      'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe',
      path.join(os.homedir(), 'AppData', 'Local', 'Google', 'Chrome', 'Application', 'chrome.exe')
    ];
    for (const p of paths) {
      if (fs.existsSync(p)) return p;
    }
  } else if (process.platform === 'darwin') {
    const p = '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';
    if (fs.existsSync(p)) return p;
  } else {
    const paths = [
      '/usr/bin/google-chrome',
      '/usr/bin/google-chrome-stable',
      '/usr/bin/chromium-browser',
      '/usr/bin/chromium'
    ];
    for (const p of paths) {
      if (fs.existsSync(p)) return p;
    }
  }
  return null;
}

/**
 * 抓取微信公众号文章
 * @param {Object} options - 配置选项
 * @param {string} options.url - 文章 URL
 * @param {boolean} options.saveToFile - 是否保存文件 (默认 false)
 * @param {string} options.outputDir - 输出目录 (默认系统临时目录)
 * @param {string} options.chromePath - Chrome 路径 (可选)
 * @param {string} options.format - 输出格式：text|markdown|html (默认 markdown)
 * @param {boolean} options.downloadImages - 是否下载图片 (默认 true)
 * @param {boolean} options.ocrImages - 是否 OCR 识别图片文字 (默认 false)
 * @returns {Promise<{title: string, author: string, time: string, content: string, images: Array}>}
 */
async function fetchWechatArticle(options = {}) {
  const {
    url,
    saveToFile = false,
    outputDir = path.join(os.tmpdir(), 'wechat-articles'),
    chromePath = findChromePath(),
    format = 'markdown',
    downloadImages = true,
    ocrImages = false
  } = options;

  if (!url) {
    throw new Error('缺少文章 URL');
  }

  if (!chromePath) {
    throw new Error('未找到 Chrome 浏览器，请手动指定 chromePath 或安装 Chrome');
  }

  let browser;
  try {
    console.log('🚀 正在启动 Chrome...');
    
    browser = await puppeteer.launch({
      executablePath: chromePath,
      headless: true,
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-gpu'
      ]
    });
    
    console.log('✅ Chrome 已启动');
    
    const page = await browser.newPage();
    
    // 移动端 User-Agent
    await page.setUserAgent('Mozilla/5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 MicroMessenger/8.0.0');
    await page.setViewport({ width: 375, height: 812 });
    
    console.log('📖 正在加载文章:', url);
    await page.goto(url, {
      waitUntil: 'networkidle2',
      timeout: 60000
    });
    
    console.log('✅ 页面加载完成');
    
    // 等待内容
    await page.waitForSelector('#js_content, .rich_media_content', { timeout: 10000 })
      .catch(() => console.log('⚠️  未找到标准内容选择器'));
    
    // 确保输出目录存在
    if (saveToFile && !fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }
    
    // 提取正文、元信息和图片
    const articleData = await page.evaluate(() => {
      // 检查是否被拦截或页面已删除
      const titleEl = document.querySelector('#activity-name') || document.querySelector('title');
      const title = titleEl?.innerText?.trim() || '无标题';
      
      const isDeleted = document.body.innerText.includes('该内容已被发布者删除') || 
                        document.body.innerText.includes('此内容因违规无法查看');
      
      if (isDeleted) {
        return { error: '文章已被删除或违规下架' };
      }

      // 作者
      const author = document.querySelector('.rich_media_meta_nickname')?.innerText?.trim() || '未知';
      
      // 时间
      const time = document.querySelector('.rich_media_meta_text')?.innerText?.trim() || '未知';
      
      // 提取图片和文字
      const contentEl = document.querySelector('#js_content') || 
                        document.querySelector('.rich_media_content') || 
                        document.querySelector('.rich_media_area_primary') || 
                        document.querySelector('article');
      
      const images = [];
      let markdownContent = '';
      let textContent = '';
      
      if (contentEl) {
        // 提取所有图片（支持多种标签和属性）
        const imgSelectors = [
          'img[data-src]',
          'img[src]',
          'mp-img[data-src]',
          'mp-img[src]',
          'weixin-img[data-src]',
          'section img',
          '.rich_media_content img'
        ];
        
        imgSelectors.forEach(selector => {
          const imgs = contentEl.querySelectorAll(selector);
          imgs.forEach(img => {
            const src = img.getAttribute('data-src') || 
                       img.getAttribute('data-original') || 
                       img.getAttribute('src');
            if (src && src.startsWith('http') && !images.includes(src)) {
              images.push(src);
            }
          });
        });
        
        // 遍历所有子节点提取文字
        const nodes = contentEl.childNodes;
        nodes.forEach(node => {
          if (node.nodeType === 1) { // Element node
            const tagName = node.tagName?.toLowerCase();
            if (tagName === 'img' || tagName === 'mp-img' || tagName === 'weixin-img') {
              const src = node.getAttribute('data-src') || 
                         node.getAttribute('data-original') || 
                         node.getAttribute('src');
              if (src && src.startsWith('http')) {
                const imgIndex = images.indexOf(src);
                if (imgIndex >= 0) {
                  markdownContent += `\n![图片 ${imgIndex + 1}](${src})\n`;
                }
              }
            } else if (tagName === 'p' || tagName === 'br' || tagName === 'div' || tagName === 'section') {
              const text = node.innerText?.trim();
              if (text && text.length > 0) {
                markdownContent += text + '\n\n';
                textContent += text + '\n';
              }
            } else {
              const text = node.innerText?.trim();
              if (text && text.length > 0) {
                markdownContent += text + '\n\n';
                textContent += text + '\n';
              }
            }
          } else if (node.nodeType === 3) { // Text node
            const text = node.textContent?.trim();
            if (text && text.length > 0) {
              markdownContent += text + '\n\n';
              textContent += text + '\n';
            }
          }
        });
        
        // 如果还是没有图片，尝试从整个页面查找
        if (images.length === 0) {
          const allImages = document.querySelectorAll('img[data-src], img[src]');
          allImages.forEach(img => {
            const src = img.getAttribute('data-src') || 
                       img.getAttribute('data-original') || 
                       img.getAttribute('src');
            if (src && src.startsWith('http') && 
                !src.includes('mmbiz.qpic.cn/mmbiz_png') && // 过滤微信默认图
                !images.includes(src)) {
              images.push(src);
            }
          });
        }
      }
      
      // 如果没有提取到内容，尝试 Fallback
      if (!markdownContent.trim()) {
        const clones = document.body.cloneNode(true);
        const removes = clones.querySelectorAll('script, style, nav, footer, .qr_code_pc, #js_profile_qrcode');
        removes.forEach(n => n.remove());
        textContent = clones.innerText.trim();
        markdownContent = textContent;
      }
      
      return { 
        title, 
        author, 
        time, 
        content: markdownContent,
        textContent,
        images,
        htmlContent: contentEl?.innerHTML || ''
      };
    });

    if (articleData.error) {
      throw new Error(`页面异常: ${articleData.error}`);
    }
    
    console.log('\n📄 ========== 文章信息 ==========\n');
    console.log('标题:', articleData.title);
    console.log('作者:', articleData.author);
    console.log('时间:', articleData.time);
    console.log('图片数量:', articleData.images?.length || 0);
    console.log('\n📝 ========== 文章内容 ==========\n');
    console.log(articleData.content.substring(0, 500) + (articleData.content.length > 500 ? '...' : ''));
    console.log('\n========== 文章结束 ==========\n');
    
    // 下载图片
    const downloadedImages = [];
    if (downloadImages && articleData.images && articleData.images.length > 0) {
      console.log(`🖼️  正在下载 ${articleData.images.length} 张图片...`);
      const imagesDir = path.join(outputDir, 'images');
      if (saveToFile && !fs.existsSync(imagesDir)) {
        fs.mkdirSync(imagesDir, { recursive: true });
      }
      
      for (let i = 0; i < articleData.images.length; i++) {
        const imgUrl = articleData.images[i];
        try {
          const ext = path.extname(new URL(imgUrl).pathname) || '.jpg';
          const filename = `image_${i + 1}${ext}`;
          
          if (saveToFile) {
            const imgPath = await downloadImage(imgUrl, imagesDir, filename);
            const imgData = {
              originalUrl: imgUrl,
              localPath: imgPath,
              filename: filename,
              ocrText: ''
            };
            
            // OCR 识别
            if (ocrImages) {
              imgData.ocrText = await recognizeImageText(imgPath, i + 1, articleData.images.length);
              if (imgData.ocrText) {
                console.log(`  📝 识别到 ${imgData.ocrText.length} 字`);
              } else {
                console.log(`  📝 未识别到文字`);
              }
            }
            
            downloadedImages.push(imgData);
            console.log(`  ✅ 已下载：${filename}`);
          } else {
            downloadedImages.push({
              originalUrl: imgUrl,
              localPath: null,
              filename: filename,
              ocrText: ''
            });
          }
        } catch (err) {
          console.log(`  ❌ 下载失败：${imgUrl} - ${err.message}`);
        }
      }
    }
    
    // 保存文件
    if (saveToFile) {
      const timestamp = Date.now();
      let filename, content;
      
      // 整理 OCR 识别结果
      let ocrContent = '';
      if (ocrImages && downloadedImages.length > 0) {
        ocrContent = '\n\n---\n\n## 🖼️ 图片文字识别结果\n\n';
        downloadedImages.forEach((img, idx) => {
          if (img.ocrText) {
            ocrContent += `### 图片 ${idx + 1}\n\n${img.ocrText}\n\n`;
          }
        });
      }
      
      if (format === 'html') {
        filename = `article-wechat-${timestamp}.html`;
        content = `<!DOCTYPE html>
<html>
<head><meta charset="utf-8"><title>${articleData.title}</title></head>
<body>
<h1>${articleData.title}</h1>
<p>作者：${articleData.author} | 时间：${articleData.time}</p>
${articleData.htmlContent}
${ocrImages ? '<h2>图片文字识别</h2>' + downloadedImages.map((img, i) => `<h3>图片 ${i + 1}</h3><p>${img.ocrText || '无文字'}</p>`).join('') : ''}
</body>
</html>`;
      } else if (format === 'markdown') {
        filename = `article-wechat-${timestamp}.md`;
        content = `# ${articleData.title}\n\n**作者**: ${articleData.author}  \n**时间**: ${articleData.time}\n\n---\n\n${articleData.content}${ocrContent}`;
      } else {
        filename = `article-wechat-${timestamp}.txt`;
        content = `标题：${articleData.title}\n作者：${articleData.author}\n时间：${articleData.time}\n\n${articleData.textContent}${ocrImages ? '\n\n---\n图片文字识别:\n' + downloadedImages.map((img, i) => `图片 ${i + 1}:\n${img.ocrText || '无文字'}`).join('\n\n') : ''}`;
      }
      
      const filepath = path.join(outputDir, filename);
      fs.writeFileSync(filepath, content, 'utf8');
      console.log(`💾 内容已保存到：${filepath}`);
      
      // 生成图片清单
      if (downloadedImages.length > 0) {
        const manifestPath = path.join(outputDir, `images-${timestamp}.json`);
        fs.writeFileSync(manifestPath, JSON.stringify(downloadedImages, null, 2), 'utf8');
        console.log(`📋 图片清单已保存到：${manifestPath}`);
        
        // 单独保存 OCR 结果
        if (ocrImages) {
          const ocrPath = path.join(outputDir, `ocr-${timestamp}.txt`);
          const ocrText = downloadedImages.map((img, i) => `=== 图片 ${i + 1} ===\n${img.ocrText || '无文字'}`).join('\n\n');
          fs.writeFileSync(ocrPath, ocrText, 'utf8');
          console.log(`📝 OCR 结果已保存到：${ocrPath}`);
        }
      }
    }
    
    await browser.close();
    console.log('🔒 浏览器已关闭');
    
    return {
      title: articleData.title,
      author: articleData.author,
      time: articleData.time,
      content: articleData.content,
      textContent: articleData.textContent,
      images: downloadedImages.length > 0 ? downloadedImages : articleData.images,
      imageCount: articleData.images?.length || 0
    };
    
  } catch (error) {
    console.error('❌ 错误:', error.message);
    console.error(error.stack);
    if (browser) await browser.close();
    throw error;
  }
}

// 命令行支持
if (require.main === module) {
  const args = process.argv.slice(2);
  
  if (args.includes('--test')) {
    require('./tests/run-tests');
  } else {
    const url = args.find(a => !a.startsWith('-'));
    const saveToFile = args.includes('--save');
    const format = args.find(a => a.startsWith('--format='))?.split('=')[1] || 'markdown';
    const downloadImages = !args.includes('--no-images');
    const ocrImages = args.includes('--ocr');
    
    if (!url) {
      console.log('用法：node fetch-wechat-article.js <文章 URL> [选项]');
      console.log('      node fetch-wechat-article.js --test  # 运行测试');
      console.log('选项：');
      console.log('  --save          保存到临时目录 (默认不保存)');
      console.log('  --format=TYPE   输出格式：text|markdown|html (默认 markdown)');
      console.log('  --no-images     不下载图片 (默认下载)');
      console.log('  --ocr           OCR 识别图片文字 (需要 --save)');
      console.log('示例：');
      console.log('  node fetch-wechat-article.js https://mp.weixin.qq.com/s/xxx');
      console.log('  node fetch-wechat-article.js https://mp.weixin.qq.com/s/xxx --save');
      console.log('  node fetch-wechat-article.js https://mp.weixin.qq.com/s/xxx --save --ocr');
      console.log('  node fetch-wechat-article.js https://mp.weixin.qq.com/s/xxx --save --format=html --ocr');
      process.exit(1);
    }
    
    fetchWechatArticle({ url, saveToFile, format, downloadImages, ocrImages })
      .then((result) => {
        console.log('✅ 抓取完成');
        console.log(`📊 统计：${result.imageCount} 张图片`);
      })
      .catch((error) => {
        console.error('❌ 抓取失败:', error.message);
        process.exit(1);
      });
  }
}

module.exports = { fetchWechatArticle };
